packages/sdk/src/types/runtime/user.ts

```diff
@@ -3,37 +3,61 @@
  * Licensed under the MIT License.
  */
 
-import { Context } from '../..';
+import { Context, GroupMask } from '../..';
+import readPath from '../../utils/readPath';
 import { InternalUser } from '../internal/user';
+import { Patchable } from '../patchable';
 
 export interface UserLike {
     id: string;
     name: string;
-
-    /**
-     * A grab bag of miscellaneous, possibly host-dependent, properties.
-     */
+    groups: number | GroupMask;
     properties: { [name: string]: string };
 }
 
 export interface UserSet {
     [id: string]: User;
 }
 
-export class User implements UserLike {
+export class User implements UserLike, Patchable<UserLike> {
     // tslint:disable:variable-name
     private _internal: InternalUser;
     /** @hidden */
     public get internal() { return this._internal; }
 
     private _name: string;
     private _properties: { [name: string]: string };
+    private _groups: GroupMask;
     // tslint:enable:variable-name
 
     public get context() { return this._context; }
     public get id() { return this._id; }
     public get name() { return this._name; }
-    /** @inheritdoc */
+
+    /**
+     * This user's group memberships. Some actors will behave differently depending on
+     * if the user is in at least one of a set of groups. See [[GroupMask]].
+     */
+    public get groups() {
+        if (!this._groups) {
+            this._groups = new GroupMask(this._context);
+            this._groups.allowDefault = false;
+            this._groups.onChanged(() => this.userChanged('groups'));
+        }
+        return this._groups;
+    }
+    public set groups(val) {
+        this._groups = val;

```

Comment:
It seems that multiple actors and users could share one GroupMask instance, but there is only one callback handler on GroupMask. If I created a GroupMask for "red team" and assigned it to multiple users and actors, only one would be notified of group changes.

You could have a private EventEmitter and let it handle all the listeners. You'd need a way to unregister a handler, but EventEmitter can do that too.

Reply:
Alternately, don't expose GroupMask to users. It can be an internal thing.

Reply:
I think I'd rather enforce single usage of GroupMask objects. Should be pretty easy to clone the mask in the property setters.

---
In reply to: [268855539](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/274#discussion_r268855539) [](ancestors = 268855539)

Reply:
I don't think that's viable. The whole point of this class is to provide a user-friendly frontend for the packed bit masks. How else would they provide group membership?

---
In reply to: [268868693](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/274#discussion_r268868693) [](ancestors = 268868693)